README
NAME
    List::AutoNumbered - Add sequential numbers to lists while creating them

    This module adds sequential numbers to lists of lists so you don't have to
    type all the numbers.  Its original use case was for adding line numbers
    to lists of testcases.  For example:

        use List::AutoNumbered;                             # line 1
        my $list = List::AutoNumbered->new(__LINE__);       # line 2
        $list->load("a")->                                  # line 3
            ("b")                                           # line 4
            ("c")                                           # line 5
            ("d");                                          # line 6

        # Now $list->arr is [ [3,"a"], [4,"b"], [5,"c"], [6,"d"] ]
        # and @$list is ( [3,"a"] ... ).

    In general, you can pass any number to the constructor.  For example:

        use List::AutoNumbered;
        use Test::More tests => 1;

        my $list = List::AutoNumbered->new;     # First entry will be number 1
        $list->load("a")->      # Yes, trailing arrow
            ("b")               # Magic!  Don"t need any more arrows!
            ("c")
            ("d");

        is_deeply($list->arr, [
            [1, "a"], [2, "b"], [3, "c"], [4, "d"]
        ]);     # Yes, it is!

METHODS
  new
    Constructor.  Basic usage options:

        $list = List::AutoNumbered->new();      # first list item is number 1
        $list = List::AutoNumbered->new($num);  # first list item is $num+1
        $list = List::AutoNumbered->new(-at => $num);   # ditto

    Each successive element will have the next number, unless you say
    otherwise (e.g., using LSKIP()).  Specifically, the first item in the list
    will be numbered one higher than the number passed to the
    "List::AutoNumbered" constructor.

    Constructor parameters are processed using Getargs::Mixed, so positional
    and named parameters are both OK.

   The "how" function
    You can give the constructor a "how" function that will make the list
    entry for a single load() or add() call:

        $list = List::AutoNumbered->new(-how => sub { @_ });
            # Jam everything together to make a flat array
        $list = List::AutoNumbered->new(41, sub { @_ });
            # Positional is OK, too.

    The "how" function is called as "how($num, @data)".  $num is the line
    number for load() calls, or "undef" for add() calls.  @data is whatever
    data you passed to "load()" or "add()".  For example, the default "how"
    function is:

        sub how {
            shift unless defined $_[0];     # add passes undef as the line number.
            [@_]                            # Wrap everything in an arrayref.
        }

    See "t/05-custom-list-entry.t" for examples of custom "how" functions.

  size
    Returns the size of the array.  Like "scalar @arr".

  last
    Returns the index of the last element in the array.  Like $#array.

  arr
    Returns a reference to the array being built.  Please do not modify this
    array directly until you are done loading it.  List::AutoNumbered may not
    work if you do.

    This can also be called by using the List::AutoNumbered object as an
    array:

        my $list = List::AutoNumbered->new...;
        foreach my $item (@$list) { ... }    # Instead of my $item (@{$list->arr})

  last_number
    Returns the current number stored by the instance.  This is the number of
    the most recently preceding new() or load() call.  This is not the number
    that will be given to the next record, since that depends on whether or
    not the next record has a skip (LSKIP()).

  load
    Push a new record with the next number on the front.  Usage:

        $instance->load(whatever args you want to push);

    Or, if the current record isn't associated with the number immediately
    after the previous record,

        $instance->load(LSKIP $n, args);

    where $n is the number of lines between this "load()" call and the last
    one.

    Returns a coderef that you can call to chain loads.  For example, this
    works:

        $instance->load(...)->(...)(...)(...) ... ;
        # You need an arrow ^^ here, but don't need any after that.

  add
    Add to the array being built, without inserting the number on the front. 
    Does increment the number and respect skips, for consistency.

    Returns the instance.

FUNCTIONS
  LSKIP
    A convenience function to create a skipper.  Prototyped as "($)" so you
    can use it conveniently with load():

        $instance->load(LSKIP 1, whatever args...);

    If you are using line numbers, the parameter to "LSKIP" should be the
    number of lines above the current line and below the last new() or load()
    call.  For example:

        my $instance = List::AutoNumbered->new(__LINE__);
        # A line
        # Another one
        $instance->load(LSKIP 2,    # two comment lines between new() and here
                        'some data');

INTERNAL PACKAGES
  List::AutoNumbered::Skipper
    This package represents a skip and is created by LSKIP().  No
    user-serviceable parts inside.

   new
    Creates a new skipper.  Parameters are for internal use only and are not
    part of the public API.

GLOBALS
  $TRACE
    (Default falsy) If truthy, print trace output.  Must be accessed directly
    unless requested on the "use" line.  Either of the following works:

        use List::AutoNumbered; $List::AutoNumbered::TRACE=1;
        use List::AutoNumbered q(*TRACE); $TRACE=1;

AUTHOR
    Christopher White, "<cxwembedded at gmail.com>"

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/cxw42/List-AutoNumbered/issues>.  I will be notified,
    and then you'll automatically be notified of progress on your bug as I
    make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc List::AutoNumbered

    You can also look for information at:

    *   MetaCPAN

        <https://metacpan.org/pod/List::AutoNumbered>

    *   CPAN Ratings

        <https://cpanratings.perl.org/d/List-AutoNumbered>

ACKNOWLEDGEMENTS
    Thanks to zdim <https://stackoverflow.com/users/4653379/zdim> for
    discussion and ideas in the Stack Overflow question
    <https://stackoverflow.com/q/50510809/2877364> that was the starting point
    for this module.

    Thanks to Dan Stewart for code contributions.

LICENSE AND COPYRIGHT
    Copyright 2019--2020 Christopher White.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published by
    the Free Software Foundation; or the Artistic License.

    See <http://dev.perl.org/licenses/> for more information.



Changes
Revision history for List-AutoNumbered

0.000009        2020-02-23  (TRIAL RELEASE)
    - Array overload: now `for(@$list)` works (#2, #3)
      (thanks to Dan Stewart)

0.000008        2019-05-11  (TRIAL RELEASE)
    - Add x_static_install=1

0.000007        2019-04-13
    - Production release of 0.000006-TRIAL
    - Should support Perl 5.6, although I haven't been able to test that.

0.000006-TRIAL  2019-04-09  (TRIAL RELEASE)
    - Added `how` argument to the constructor (GH-1)
    - Added dependency on Getargs::Mixed.

0.000005        2019-04-08
    Production release of 0.000004-TRIAL

0.000004-TRIAL  2019-04-04  (TRIAL RELEASE)
    Updated Makefile.PL to handle test_requires on older EUMMs

0.000003        2019-04-03
    No changes since trial version 0.000002

0.000002-TRIAL  2019-04-02  (TRIAL RELEASE)
    Added missing dependency to META

0.000001-TRIAL  2019-04-02  (TRIAL RELEASE)
    First version, released on an unsuspecting world.



support/readme.pl
#!perl
# readme_md.pl: Make README.md from a Perl file.

use 5.014;
use strict;
use warnings;
use Getopt::Long qw(:config gnu_getopt);
use Path::Class;

# Parse command-line options
my ($source_fn, $dest_fn, $appveyor, $appveyor_badge, $skipfrom, $skipto);
my $format = 'md';
GetOptions( "i|input=s" => \$source_fn,
            "o|output=s" => \$dest_fn,
            "f|format=s" => \$format,
            "appveyor=s" => \$appveyor,         # username/repo
            "avbadge=s" => \$appveyor_badge,    # default $appveyor
            "skipfrom=s" => \$skipfrom,
            "skipto=s" => \$skipto,
)
    or die "Error in arguments.  Usage:\nreadme_md.pl -i input -o output [-f format]\nFormat = md (default) or text.";

die "Need an input file" unless $source_fn;
die "Need an output file" unless $dest_fn;

die "skipfrom and skipto must be used together" if (!!$skipfrom) ^ (!!$skipto);

$appveyor =~ m{^[A-Za-z0-9-]+/[A-Za-z0-9-]+} or die '--appveyor <GH username>/<GH repo>' if $appveyor;
$appveyor_badge //= $appveyor;
$appveyor_badge =~ m{^[A-Za-z0-9-]+/[A-Za-z0-9-]+} or die '--appveyor <GH username>/<GH repo>' if $appveyor_badge;

# Load the right parser
my $parser;
if($format eq 'md') {
    require Pod::Markdown;
    $parser = Pod::Markdown->new;

} elsif($format eq 'text') {
    require Pod::Text;
    $parser = Pod::Text->new(sentence => 1, width => 78);

} else {
    die "Invalid format $format (I understand 'md' and 'text')"
}

# Turn the POD into the output format
my $parsed = '';
$parser->output_string(\$parsed);
my $pod = file($source_fn)->slurp;
$parser->parse_string_document($pod);
open my $fh, '<', \$parsed;

# Filter and tweak the POD
my $saw_name = 0;
my $tweak_name = ($format eq 'md');
my $force_conventions = ($format eq 'md');
my $output = '';

while(my $line = <$fh>) {

    # In Markdown, turn NAME into the text, as a heading.
    # Also add the Appveyor badge.
    if($tweak_name && !$saw_name && $line =~ /NAME/) {
        $saw_name = 1;
        next;
    } elsif($tweak_name && $saw_name && $line =~ m{\H\h*$/}) {
        $output .= ($format eq 'md' ? '# ' : '') . "$line\n";
        $output .= "[![Appveyor Badge](https://ci.appveyor.com/api/projects/status/github/${appveyor_badge}?svg=true)](https://ci.appveyor.com/project/${appveyor})\n\n" if $appveyor;
        $saw_name = 0;
        next;
    } elsif($tweak_name && $saw_name) {
        next;   # Waiting for the name line to come around
    }

    next if $line =~ /SYNOPSIS/;    # Don't need this header.

    # Skip the internals, if any
    if($skipfrom) {
        $output .= $line if $line =~ /$skipto/;
        next if ($line =~ /$skipfrom/)..($line =~ /$skipto/);
    }

    $output .= $line;   # Copy everything that's left.
}

file($dest_fn)->spew($output);

__END__
# Documentation ========================================================== {{{1

=head1 NAME

readme.pl - Generate README/README.md from POD.

=head1 SYNOPSIS

    readme.pl -i <input filename> -o <output filename> [args...]

=head1 ARGUMENTS

=head2 B<-i>, B<--input>

(Required) The input filename.  Must be something containing POD.

=head2 B<-o>, B<--output>

(Required) The output filename.

=head2 B<-f>, B<--format>

(Optional, default C<md>) The format to generate.  Values are C<md> and C<text>.

=head2 B<--appveyor> <slug>

Add an appveyor badge.  B<< <slug> >> should be of the form C<username/repo>,
and should point to the build page.

=head2 B<--avbadge> <slug>

Specify the badge image URL, if different from the C<--appveyor> location.
B<< <slug> >> should be of the form C<username/repo>.

=head2 B<--skipfrom> <line1> B<--skipto> <line2>

Must be used together or not at all.  If specified, lines starting at
B<< <line1> >> and extending to just B<before> B<< <line2> >> will be omitted
from the output.  For example, if your POD has an C<INTERNALS> section you
want to leave out of the readme, and the next section is C<AUTHOR>, specify

    --skipfrom INTERNALS --skipto AUTHOR

=head1 LICENSE

The same terms as Perl itself.  NO WARRANTY.

=cut

# }}}1
# vi: set fdm=marker: #


lib/List/AutoNumbered.pm
package List::AutoNumbered;

use 5.006;
use strict;
use warnings;
use overload
    '@{}' => \&arr;

our $VERSION = '0.000009';  # TRIAL

# Exports
use parent 'Exporter';
our (@EXPORT, @EXPORT_OK, %EXPORT_TAGS);
BEGIN {
    @EXPORT = qw(LSKIP);
    @EXPORT_OK = qw(*TRACE);    # can be localized
    %EXPORT_TAGS = (
        default => [@EXPORT],
        all => [@EXPORT, @EXPORT_OK]
    );
}

# Imports
use Getargs::Mixed;

# Documentation ======================================================== {{{1

=head1 NAME

List::AutoNumbered - Add sequential numbers to lists while creating them

=head1 SYNOPSIS

This module adds sequential numbers to lists of lists so you don't have to
type all the numbers.  Its original use case was for adding line numbers
to lists of testcases.  For example:

    use List::AutoNumbered;                             # line 1
    my $list = List::AutoNumbered->new(__LINE__);       # line 2
    $list->load("a")->                                  # line 3
        ("b")                                           # line 4
        ("c")                                           # line 5
        ("d");                                          # line 6

    # Now $list->arr is [ [3,"a"], [4,"b"], [5,"c"], [6,"d"] ]
    # and @$list is ( [3,"a"] ... ).

In general, you can pass any number to the constructor.  For example:

    use List::AutoNumbered;
    use Test::More tests => 1;

    my $list = List::AutoNumbered->new;     # First entry will be number 1
    $list->load("a")->      # Yes, trailing arrow
        ("b")               # Magic!  Don"t need any more arrows!
        ("c")
        ("d");

    is_deeply($list->arr, [
        [1, "a"], [2, "b"], [3, "c"], [4, "d"]
    ]);     # Yes, it is!

=cut

our $TRACE = 0; # Documented below

=head1 METHODS

=cut

# }}}1

# Internal helpers {{{1

# Defined-or
sub _dor { (defined $_[0]) ? $_[0] : $_[1] }

# }}}1

=head2 new

Constructor.  Basic usage options:

    $list = List::AutoNumbered->new();      # first list item is number 1
    $list = List::AutoNumbered->new($num);  # first list item is $num+1
    $list = List::AutoNumbered->new(-at => $num);   # ditto

Each successive element
will have the next number, unless you say otherwise (e.g., using
L<LSKIP()|/LSKIP>).  Specifically, the first item in the list will be numbered
one higher than the number passed to the C<List::AutoNumbered> constructor.

Constructor parameters are processed using L<Getargs::Mixed>, so positional
and named parameters are both OK.

=head3 The C<how> function

You can give the constructor a "how" function that will make the list entry
for a single L<load()|/load> or L<add()|/add> call:

    $list = List::AutoNumbered->new(-how => sub { @_ });
        # Jam everything together to make a flat array
    $list = List::AutoNumbered->new(41, sub { @_ });
        # Positional is OK, too.

The C<how> function is called as C<how($num, @data)>.  C<$num> is the
line number for L<load()|/load> calls, or C<undef> for L<add()|/add> calls.
C<@data> is whatever data you passed to C<load()> or C<add()>.  For example,
the default C<how> function is:

    sub how {
        shift unless defined $_[0];     # add passes undef as the line number.
        [@_]                            # Wrap everything in an arrayref.
    }

See C<t/05-custom-list-entry.t> for examples of custom C<how> functions.

=cut

sub new {
    my ($class, %args) = parameters('self',[qw(; at how)], @_);
    my $self = bless {
        num => _dor($args{at}, 0),  # The last number used
        arr => [],                  # The data
        how => _dor($args{how},     # Routine to make a list item
                    sub {
                        shift unless defined $_[0];     # add passes undef
                                                        # as the line number.
                        [@_]        # By default, just wrap it
                    }
                ),
    }, $class;

    # Make a loader that adds an item and returns itself --- not $self.
    # Note that $self is captured --- the loader function does not take
    # a $self argument.
    $self->{loader} = sub { $self->_L(@_); return $self->{loader} };

    print "# Created - now $self->{num}\n" if $TRACE;
    return $self;
} #new()

# Accessors {{{1

=head2 size

Returns the size of the array.  Like C<scalar @arr>.

=cut

sub size { scalar @{ shift->{arr} }; }

=head2 last

Returns the index of the last element in the array.  Like C<$#array>.

=cut

sub last { shift->size-1; }

=head2 arr

Returns a reference to the array being built.  Please do not modify this
array directly until you are done loading it.  List::AutoNumbered may not
work if you do.

This can also be called by using the List::AutoNumbered object as an array:

    my $list = List::AutoNumbered->new...;
    foreach my $item (@$list) { ... }    # Instead of my $item (@{$list->arr})

=cut

sub arr { shift->{arr}; }

=head2 last_number

Returns the current number stored by the instance.  This is the number
of the most recently preceding L<new()|/new> or L<load()|/load> call.
This is B<not> the number that will be given to the next record, since that
depends on whether or not the next record has a skip (L<LSKIP()|/LSKIP>).

=cut

sub last_number { shift->{num} }

# }}}1
# Loading {{{1

=head2 load

Push a new record with the next number on the front.  Usage:

    $instance->load(whatever args you want to push);

Or, if the current record isn't associated with the number immediately after
the previous record,

    $instance->load(LSKIP $n, args);

where C<$n> is the number of lines between this C<load()> call and the last one.

Returns a coderef that you can call to chain loads.  For example, this works:

    $instance->load(...)->(...)(...)(...) ... ;
    # You need an arrow ^^ here, but don't need any after that.

=cut

sub load { goto &{ shift->{loader} } }  # kick off loading

# _L: Implementation of load()
sub _L {
    my $self = shift;

    shift if $self->_update_lnum(@_);   # Check for skipped lines from LSKIP()

    push @{ $self->{arr} }, $self->{how}->($self->{num}, @_);
    return $self;
} #_L()

=head2 add

Add to the array being built, B<without> inserting the number on the front.
Does increment the number and respect skips, for consistency.

Returns the instance.

=cut

sub add {   # just add it
    my $self = shift;
    shift if $self->_update_lnum(@_);   # Check for skipped lines from LSKIP()
    push @{ $self->{arr} }, $self->{how}->(undef, @_);
    return $self;
} #add

# }}}1
# Skipping {{{1

=head1 FUNCTIONS

=head2 LSKIP

A convenience function to create a skipper.  Prototyped as C<($)> so you can
use it conveniently with L<load()|/load>:

    $instance->load(LSKIP 1, whatever args...);

If you are using line numbers, the parameter to C<LSKIP> should be the number
of lines above the current line and below the last L<new()|/new> or
L<load()|/load> call.  For example:

    my $instance = List::AutoNumbered->new(__LINE__);
    # A line
    # Another one
    $instance->load(LSKIP 2,    # two comment lines between new() and here
                    'some data');

=cut

sub LSKIP ($) {
    List::AutoNumbered::Skipper->new(@_);
} #LSKIP()

# _update_lnum: Increment the line number, and run a skip if there is one.
# Call from a method as:
#   my $self = shift;
#   shift if $self->_update_lnum(@_);

sub _update_lnum {
    my $self = shift;

    if(ref $_[0] eq 'List::AutoNumbered::Skipper') {    # implies scalar @_
        $self->{num} += $_[0]->{how_many} + 1;
        print "# Skipped $_[0]->{how_many} - now $self->{num}\n" if $TRACE;
        return 1;   # We do need to shift

    } else {
        ++$self->{num};
        print "# Incremented - now $self->{num}\n" if $TRACE;
        return 0;       # No skipper, so don't shift it off.
    }
} #_update_lnum()

=head1 INTERNAL PACKAGES

=head2 List::AutoNumbered::Skipper

This package represents a skip and is created by L<LSKIP()|/LSKIP>.
No user-serviceable parts inside.

=cut

{
    package List::AutoNumbered::Skipper;
    use Scalar::Util qw(looks_like_number);

=head3 new

Creates a new skipper.  Parameters are for internal use only and are not
part of the public API.

=cut

    sub new {
        my $class = shift;
        die "Need a single number" unless @_==1 and looks_like_number $_[0];
        bless {how_many => $_[0]}, $class;
    }
} #List::AutoNumbered::Skipper

# }}}1

1; # End of List::AutoNumbered

# Rest of the docs {{{1
__END__

=head1 GLOBALS

=head2 $TRACE

(Default falsy) If truthy, print trace output.  Must be accessed directly
unless requested on the C<use> line.  Either of the following works:

    use List::AutoNumbered; $List::AutoNumbered::TRACE=1;
    use List::AutoNumbered q(*TRACE); $TRACE=1;

=head1 AUTHOR

Christopher White, C<< <cxwembedded at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/cxw42/List-AutoNumbered/issues>.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc List::AutoNumbered

You can also look for information at:

=over 4

=item * MetaCPAN

L<https://metacpan.org/pod/List::AutoNumbered>

=item * Appveyor (CI builds)

L<https://ci.appveyor.com/project/cxw42/list-autonumbered>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to L<zdim|https://stackoverflow.com/users/4653379/zdim>
for discussion and ideas in the
L<Stack Overflow question|https://stackoverflow.com/q/50510809/2877364>
that was the starting point for this module.

Thanks to Dan Stewart for code contributions.

=head1 LICENSE AND COPYRIGHT

Copyright 2019--2020 Christopher White.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

# }}}1
# vi: set fdm=marker: #


lib/List/LineNumbered.pm
package List::LineNumbered;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.000001';  # TRIAL

use parent 'Exporter';
our @EXPORT = qw(LSKIP);

# Documentation ======================================================== {{{1

=head1 NAME

List::LineNumbered - Add line numbers to lists while creating them

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use List::LineNumbered;

    my $foo = List::LineNumbered->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=cut

# }}}1

# Internals {{{1

# Defined-or
sub _dor { (defined $_[0]) ? $_[0] : $_[1] }

# }}}1

=head1 new

Constructor.  Call as $class->new(__LINE__); each element is one line.

=cut
sub new {
    my $class = shift;
    my $self = bless {lnum => _dor(shift, 0), arr => []}, $class;

    # Make a loader that adds an item and returns itself --- not $self
    $self->{loader} = sub { $self->L(@_); return $self->{loader} };
        # TODO add a skip() method callable on the loader

    return $self;
} #new()

# Accessors {{{1
sub size { return scalar @{ shift->{arr} }; }
sub last { return shift->size-1; }      # $#
sub arr { return shift->{arr}; }

# }}}1
# Loading {{{1
sub load { goto &{ shift->{loader} } }  # kick off loading

=head2 L

Push a new record with the next line number on the front.  Usage:

    $instance->L(whatever args you want to push);

Or, if the current record isn't on the line immediately after the previous
record,

    $instance->L(

=cut

sub L {
    my $self = shift;

    # Check for skipped lines from LSKIP()
    if(@_ && ref $_[0] eq 'TestcaseList::Skipper') {
        $self->{lnum} += $_[0]->{how_many};
        shift;
    }

    push @{ $self->{arr} }, [++$self->{lnum}, @_];
    return $self;
} #L

sub add {   # just add it
    my $self = shift;
    ++$self->{lnum};    # keep it consistent
    push @{ $self->{arr} }, [@_];
    return $self;
} #add

# }}}1
# Skipping {{{1

=head2 LSKIP

A convenience function to create a skipper.  Prototyped as C<($)> so you can
use it conveniently with L</L>:

    $instance->L(LSKIP 1, whatever args...);

=cut

sub LSKIP ($) {
    TestcaseList::Skipper->new(@_);
} #LSKIP()

{
    package TestcaseList::Skipper;
    use Scalar::Util qw(looks_like_number);

    sub new {
        my $class = shift;
        die "Need a number" unless @_==1 and looks_like_number $_[0];
        bless {how_many => $_[0]}, $class;
    }
}

# }}}1

1; # End of List::LineNumbered

# Rest of the docs {{{1
__END__

=head1 AUTHOR

Christopher White, C<< <cxwembedded at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/cxw42/List-LineNumbered/issues>.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc List::LineNumbered

You can also look for information at:

=over 4

=item * MetaCPAN

L<https://metacpan.org/pod/List::LineNumbered>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/List-LineNumbered>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to zdim for discussion on the
L<Stack Overflow question|https://stackoverflow.com/q/50510809/2877364>
that was the starting point for this module.

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Christopher White.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

# }}}1
# vi: set fdm=marker: #


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::Spec;
use Config;

my $VERSION_FROM = File::Spec->catfile(qw(lib List AutoNumbered.pm));
my $IS_TRIAL = check_trial();
my $secure_perl_path = get_perl_filename();
my @provides = ();

# Check if this is a TRIAL version {{{1
sub check_trial {
    TRIAL: {
        # Get the VERSION line
        open my $fd, '<', $VERSION_FROM or last TRIAL;
        my $linetext;
        while(<$fd>) {
            next unless /VERSION/;
            $linetext = $_;
            last;
        }
        close $fd;
        return !!($linetext =~ /\bTRIAL\b/);
    }
    return 0;
} #check_trial()

# }}}1
# Module metadata {{{1
eval {
    require Module::Metadata;
    @provides =
        (provides => Module::Metadata->provides(version => '2', dir => 'lib'));
        # Thanks to https://stackoverflow.com/a/28928985/2877364 by LEONT
        # for suggesting Module::Metadata.
};

# }}}1
# Get the filename of the Perl interpreter running this. {{{1
# Modified from perlvar.
# The -x test is for cygwin or other systems where $Config{perlpath} has no
# extension and $Config{_exe} is nonempty.  E.g., symlink perl->perl5.10.1.exe.
# There is no "perl.exe" on such a system.
sub get_perl_filename {
    my $secure_perl_path = $Config{perlpath};
    if ($^O ne 'VMS') {
        $secure_perl_path .= $Config{_exe}
            unless (-x $secure_perl_path) ||
                            ($secure_perl_path =~ m/$Config{_exe}$/i);
    }
    return $secure_perl_path;
} # get_perl_filename()

# }}}1
# Makefile customization (MY) {{{1
{   package MY;

    # dist_core: make `dist` a :: target rather than a : target,
    # and add distcheck before dist.  Also, add -TRIAL to the tgz if necessary.
    sub dist_core {
        my $self = shift;
        my $text = $self->SUPER::dist_core(@_);
        $text =~ s/^dist\s*:[^:]/dist:: distcheck /m;

        # Add -TRIAL if it's a trial release
        if($IS_TRIAL) {
            # Note: we don't have to worry about EOL; Appveyor uses gmake even
            # on Windows, and it only uses \n.
            print STDERR "TRIAL version\n";
            my $newtext .=  # Command to rename the tgz.  TODO see if this works on Windows.
                "\t\"$secure_perl_path\" -MExtUtils::Command -e mv -- " .
                '"$(DISTVNAME).tar$(SUFFIX)" ' .
                '"$(DISTVNAME)-TRIAL.tar$(SUFFIX)"' . "\n";

            # Insert $newtext at the end of the `dist` target
            $text =~ s{
                ^(dist\h*:.*\n      # dist header line
                    (?:\h+\S.*\n)+) # dist body lines.  `.` doesn't match `\n`.
                                    # NOTE: on Appveyor, the continuation line
                                    # begins with a space rather than a tab.
                                    # Therefore, look for \h after \n.
                                    # Not \s, because that matches a \n!
            }{$1$newtext}mx;
        } #endif $IS_TRIAL

        return $text;
    } # MY::dist_core

    # Generate README.md; add uninstall target and other test targets
    sub postamble {     # TODO also handle Windows nmake syntax (SET vs. export)

        # Note: pod2markdown is marked with a hyphen so that `make` can
        # succeed and generate META files even if Pod::Markdown is not yet
        # installed.

        my $make_readme_md = File::Spec->catfile(qw(support readme.pl));
        return <<EOT;
authortest:
\tRELEASE_TESTING=1 prove -lvj4 xt

testhere:   # Run the tests from lib rather than blib
\tprove -lj4

README.md: @{[$VERSION_FROM]} Makefile.PL $make_readme_md
\t-"$secure_perl_path" "$make_readme_md" -i "\$<" -o "\$@" -f md

README: @{[$VERSION_FROM]} Makefile.PL $make_readme_md
\t-"$secure_perl_path" "$make_readme_md" -i "\$<" -o "\$@" -f text

all :: README.md README
dist :: README.md README

# Uninstallation: mimic what the normal "uninstall" target does, but actually
# carry out the actions.
# Note: this may leave empty directories in place, since the packlist only
# includes files.

ACTUALLY_UNINSTALL = "$secure_perl_path" -MExtUtils::Install -e 'uninstall(shift,1)' --

actually_uninstall :: actually_uninstall_from_\$(INSTALLDIRS)dirs
\t\$(NOECHO) \$(NOOP)

actually_uninstall_from_perldirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(PERL_ARCHLIB)/auto/\$(FULLEXT)/.packlist"

actually_uninstall_from_sitedirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(SITEARCHEXP)/auto/\$(FULLEXT)/.packlist"

actually_uninstall_from_vendordirs ::
\t-\$(ACTUALLY_UNINSTALL) "\$(VENDORARCHEXP)/auto/\$(FULLEXT)/.packlist"

EOT

    } #postamble()
} #package MY

# }}}1

# Main options for EUMM
my %opts = (
    NAME             => 'List::AutoNumbered',
    AUTHOR           => q{Christopher White <cxwembedded@gmail.com>},
    VERSION_FROM     => $VERSION_FROM,
    ABSTRACT_FROM    => $VERSION_FROM,
    LICENSE          => 'perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.006',
    CONFIGURE_REQUIRES => {
        'Config' => '0',
        'ExtUtils::MakeMaker' => '0',
        'File::Spec' => '0',
    },
    TEST_REQUIRES => {
        'Data::Dumper' => '0',  # to give us a random blessed reference
        'lib::relative' => '0.002',
        'Import::Into' => '0',
        'Test::Fatal' => '0',
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'Getargs::Mixed' => '1.06', # Supports perl 5.6
        'Exporter' => '0',
        'parent' => '0',
        'Scalar::Util' => '1.50',
        'strict' => '0',
        'warnings' => '0',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'List-AutoNumbered-*' },

    META_MERGE => {
        'meta-spec' => { version => 2 },

        resources => {
            bugtracker => {
                web  => 'https://github.com/cxw42/List-AutoNumbered/issues',
            },
            repository => {
                type => 'git',
                url  => 'https://github.com/cxw42/List-AutoNumbered.git',
                web  => 'https://github.com/cxw42/List-AutoNumbered',
            },
        },

        prereqs => {
            develop => {
                requires => {
                    'App::RewriteVersion' => '0',   # for perl-bump-version
                    'Module::Metadata' => '1.000016',
                    'Path::Class' => '0',           # for support/readme.pl
                },
            },
        },

        # The gymnastics in this file are primarily for developer
        # convenience.  The actual dist, if processed by an installer
        # new enough to support x_static_install, doesn't need any of
        # the above processing for a plain installation.  Therefore,
        # x_static_install is OK.  For more, see
        # http://blogs.perl.org/users/shoichi_kaji1/2017/03/make-your-cpan-module-static-installable.html
        "x_static_install" => "1",

        @provides,

    }, #META_MERGE
);

$opts{META_MERGE}->{release_status} = 'testing' if $IS_TRIAL;

# Some tests run on older versions of EUMM that do not support TEST_REQUIRES.
# http://www.cpantesters.org/cpan/report/f445404a-56b8-11e9-bab0-5e087347484a
# is one example.  This is a workaround, modified from
# https://metacpan.org/source/RJBS/JSON-Typist-0.005/Makefile.PL ,
# linked from http://blogs.perl.org/users/neilb/2017/05/specifying-dependencies-for-your-cpan-distribution.html .
# If we are on a version of EUMM that doesn't support TEST_REQUIRES,
# move those dependencies to PREREQ_PM.
unless( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
    $opts{PREREQ_PM}->{$_} = $opts{TEST_REQUIRES}->{$_}
        foreach keys %{$opts{TEST_REQUIRES}};
    delete $opts{TEST_REQUIRES};
}

WriteMakefile(%opts);
# vi: set fdm=marker: #
